### PR TITLE
Repair Command

### LINKED ISSUE #327: [8pt] Repair: Base
Sometimes we have to repair cluster using brute force - by manually editing `.config.yml`.
It could be useful to automate this flow.

We need to support these actions:

* change instance advertise URI;
* remove node from cluster;
* change replicaset master.

### PR DESCRIPTION
**Pull request by dokshina**
Sometimes we want to patch topology config by hands. 
`cartridge repair` command allows to do that.

I didn't forget about

- [x] Tests
- [ ] Changelog
- [x] Documentation

Closes #327

### CODE REVIEW

File: cli/commands/usage.go [hunk 1]
```diff
@@ -105,7 +105,7 @@ Defaults to ./tmp/run (or "run-dir" in .cartridge.yml)
 
 	dataDirUsage = `Directory where instances' data is stored
```
Comment (olegrok) [inline cli/commands/usage.go:106]: `instances'` -> `instance's`

File: cli/commands/usage.go [hunk 2]
```diff
@@ -136,6 +136,21 @@ If specified, "INSTANCE_NAME..." are ignored.
 `
 )
 
+// REPAIR
+const (
+	advertiseURIUsage = `Change instance advertise URI
+`
+
+	repairDataDirUsage = `Directory where instances' data is stored
```
Comment (olegrok) [inline cli/commands/usage.go:144]: `instances'` -> `instance's`

File: cli/common/result.go [hunk 3]
```diff
@@ -0,0 +1,63 @@
+package common
```
Comment (olegrok) [inline cli/common/result.go:1]: Could be easily moved into separate patch

Comment (dokshina) [reply cli/common/result.go:1]: Yes, I agree. I'm thinking about moving it to separated module `result`, what do you think?

Comment (olegrok) [reply cli/common/result.go:1]: Not sure about module. Why is simple file not enough?

Comment (dokshina) [reply cli/common/result.go:1]: See #331

File: README.rst [hunk 4]
```diff
@@ -127,7 +127,8 @@ The following commands are supported:
 * ``status`` — get current instance(s) status;
 * ``log`` — get logs of instance(s);
 * ``clean`` - clean instance(s) files;
-* ``pack`` — pack the application into a distributable bundle.
+* ``pack`` — pack the application into a distributable bundle;
+* ``repair`` — repair the cluster.
```
Comment (olegrok) [inline README.rst:131]: I believe it should be extended. It's not clear that means "repair the cluster"

File: cli/commands/usage.go [hunk 5]
```diff
@@ -137,6 +137,20 @@ If specified, "INSTANCE_NAME..." are ignored.
...
+const (
+	advertiseURIUsage = `Change instance advertise URI
+`
+
+	repairDataDirUsage = `Directory where instances data is stored
+Defaults to /var/lib/tarantool
+`
+
+	dryRunUsage = `Run command in dry-run mode
+Show changes but don't apply them
```
Comment (olegrok) [inline cli/commands/usage.go:150]: `don't` -> `doesn't`

File: cli/common/utils.go [hunk 6]
```diff
@@ -237,3 +237,45 @@ Loop:
 
 	return lastNewLinePos, nil
 }
+
+func ConvertToStringsSlice(s interface{}) ([]string, error) {
+	interfacesSlice, ok := s.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("Isn' t a list")
```
Comment (olegrok) [inline cli/common/utils.go:244]: `Isn' t` -> `Isn't` or even '<smth> is not a list'

File: cli/common/utils.go [hunk 7]
```diff
@@ -237,3 +237,45 @@ Loop:
 
 	return lastNewLinePos, nil
 }
+
+func ConvertToStringsSlice(s interface{}) ([]string, error) {
```
Comment (olegrok) [inline cli/common/utils.go:241]: Could it be declared as `s []interface{}`?

File: cli/common/utils.go [hunk 8]
```diff
@@ -237,3 +237,45 @@ Loop:
...
+	interfacesSlice, ok := s.([]interface{})
+	if !ok {
+		return nil, fmt.Errorf("Isn' t a list")
+	}
+
+	stringsSlice := make([]string, len(interfacesSlice))
+	for i, elem := range interfacesSlice {
+		stringElem, ok := elem.(string)
+		if !ok {
+			return nil, fmt.Errorf("Slice element %d isn't a string", i)
```
Comment (olegrok) [inline cli/common/utils.go:251]: And add "%t" for such element

File: cli/repair/advertise_uri.go [hunk 9]
```diff
@@ -0,0 +1,25 @@
+package repair
+
+import (
+	"fmt"
+
+	"github.com/tarantool/cartridge-cli/cli/context"
+)
+
+func patchConfAdvertiseURI(workDir string, ctx *context.Ctx) ([]string, error) {
```
Comment (olegrok) [inline cli/repair/advertise_uri.go:9]: Is it really needed?
This function just call another function.

File: cli/repair/clusterwide_config.go [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/tarantool/cartridge-cli/cli/common"
+	"gopkg.in/yaml.v2"
+)
+
+const (
+	configDirName        = "config"
```
Comment (olegrok) [inline cli/repair/clusterwide_config.go:13]: This rattling mix of aligns can kill any perfectionist

File: cli/repair/clusterwide_config.go [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+	instancesConfRawMap, ok := instancesConfRaw.(RawConfType)
+	if !ok {
+		return nil, fmt.Errorf("%q value isn't a map", keyInstances)
+	}
+
+	topologyConf.InstancesRaw = instancesConfRawMap
+
+	instancesConf := make(map[string]InstanceConfType)
+
+	for instanceUUIDRaw, instanceConfRaw := range instancesConfRawMap {
```
Comment (olegrok) [inline cli/repair/clusterwide_config.go:124]: Do you validate correctness of specified UUID/URI? Btw null uuid (00000000-0000-0000-0000-000000000000) should be banned as well.

File: cli/repair/clusterwide_config.go [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+	return &instancesConf, nil
+}
+
+func setInstanceURIRaw(topologyConf *TopologyConfType, instanceUUID, newURI string) error {
+	instanceConf, ok := topologyConf.Instances[instanceUUID]
+	if !ok {
+		return fmt.Errorf("Instance %s isn't found in cluster", instanceUUID)
+	}
+
+	if instanceConf.IsExpelled {
```
Comment (olegrok) [inline cli/repair/clusterwide_config.go:183]: I this we should introduce status name, not flags "isExpelled"/"isDisabled" etc. It allows to add new statuses when it will be introduced.

File: cli/repair/clusterwide_config.go [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+	return nil
+}
+
+func removeInstanceFromRaw(topologyConf *TopologyConfType, instanceUUID string) {
+	delete(topologyConf.InstancesRaw, instanceUUID)
+}
+
+// REPLICASETS
+
+func getReplicasetsConf(topologyConf *TopologyConfType) (*map[string]ReplicasetConfType, error) {
```
Comment (olegrok) [inline cli/repair/clusterwide_config.go:198]: It's too hard to understand code then validation logic is mixed with business logic. Please split them for all cases.

File: cli/repair/repair.go [hunk 14]
```diff
@@ -0,0 +1,95 @@
...
+	for i := 0; i < len(appWorkDirNames); i++ {
+		select {
+		case res := <-resCh:
+			log.Infof(res.String())
+			if res.Status != common.ResStatusOk {
+				errors = append(errors, res.FormatError())
+			}
+
+			if ctx.Cli.Verbose {
+				for _, message := range res.Messages {
```
Comment (olegrok) [inline cli/repair/repair.go:80]: I think we need to provide some title for following message list

File: cli/repair/repair.go [hunk 15]
```diff
@@ -0,0 +1,95 @@
...
+			resCh <- res
+		}(workDirPath, workDirName, resCh)
+	}
+
+	var errors []error
+
+	for i := 0; i < len(appWorkDirNames); i++ {
+		select {
+		case res := <-resCh:
+			log.Infof(res.String())
```
Comment (olegrok) [inline cli/repair/repair.go:74]: What does usually this command print?

File: test/clusterwide_conf.py [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+def assert_conf_changed(conf_paths, other_app_conf_paths, old_conf, new_conf):
+    for conf_path in conf_paths:
+        assert os.path.exists(conf_path)
+
+        with open(conf_path, 'r') as f:
+            conf = yaml.safe_load(f.read())
+            assert conf == new_conf
+
+        # check backup
+        backup_conf_path = '%s.bak' % conf_path
```
Comment (olegrok) [inline test/clusterwide_conf.py:64]: maybe `.backup`?

File: cli/repair/clusterwide_config.go (hunk 12 above)
Comment (dokshina) [reply cli/repair/clusterwide_config.go:183]: This statuses are processed in a very different way, I don't want to compare status with a constant, it's easier to use flag.

File: cli/repair/clusterwide_config.go (hunk 13 above)
Comment (dokshina) [reply cli/repair/clusterwide_config.go:198]: You mean field types validation? I don't see any reasons to separate them from type casts.

`clusterwide_config` functions hide raw config details from business logic, and I sure it's enough.

File: cli/repair/repair.go (hunk 15 above)
Comment (dokshina) [reply cli/repair/repair.go:74]: `myapp.instance-1: OK`

File: cli/common/utils.go (hunk 7 above)
Comment (dokshina) [reply cli/common/utils.go:241]: No, since it should be converted `interface{} -> []interface{} -> []string`

File: README.rst (hunk 4 above)
Comment (dokshina) [reply README.rst:131]: What about `Patch cluster-wide config file`?

File: cli/repair/clusterwide_config.go (hunk 10 above)
Comment (dokshina) [reply cli/repair/clusterwide_config.go:13]: It's go formatter.

File: cli/repair/advertise_uri.go (hunk 9 above)
Comment (dokshina) [reply cli/repair/advertise_uri.go:9]: Yes, patch functions call `patchConf` wrapper and can be used by `repair.Run` function.

Look at `patchConfRemoveInstance`.

File: cli/repair/clusterwide_config.go (hunk 11 above)
Comment (dokshina) [reply cli/repair/clusterwide_config.go:124]: It's not `cartridge-cli`'s business. It's about clusterwide config format.

File: README.rst (hunk 4 above)
Comment (olegrok) [reply README.rst:131]: "Patch cluster configuration files"

File: cli/repair/repair.go (hunk 15 above)
Comment (olegrok) [reply cli/repair/repair.go:74]: ok

File: cli/repair/repair.go (hunk 14 above)
Comment (dokshina) [reply cli/repair/repair.go:80]: For example, what?

File: cli/commands/repair.go [hunk 17]
```diff
@@ -0,0 +1,136 @@
...
+func runRepairListCommand(cmd *cobra.Command, args []string) error {
+	if ctx.Project.Name == "" {
+		return fmt.Errorf("Please, specify application name using --name")
+	}
+
+	return repair.List(&ctx)
+}
+
+func runRepairURICommand(cmd *cobra.Command, args []string) error {
+	if ctx.Project.Name == "" {
```
Comment (olegrok) [inline cli/commands/repair.go:107]: Such check is in three places. Could you move it to single place?

File: cli/repair/clusterwide_config.go [hunk 18]
```diff
@@ -0,0 +1,314 @@
...
+)
+
+const (
+	configDirName        = "config"
+	topologyConfFilename = "topology.yml"
+
+	keyInstances            = "servers"
+	keyReplicasets          = "replicasets"
+	keyInstanceAdvertiseURI = "uri"
+	keyInstanceDisabled     = "disabled"
```
Comment (olegrok) [inline cli/repair/clusterwide_config.go:20]: Why is there no `keyInstanceExpelled` or `disabledState `?

File: cli/repair/clusterwide_config.go (hunk 11 above)
Comment (olegrok) [reply cli/repair/clusterwide_config.go:119]: You give an instrument to patch configuration by hands. You should prevent such invalid/dangerous cases

File: cli/repair/common.go [hunk 19]
```diff
@@ -0,0 +1,180 @@
...
+	if _, err := os.Stat(path); err != nil {
+		return fmt.Errorf("Failed to use specified path: %s", err)
+	}
+
+	backupPath := getBackupPath(path)
+	if _, err := os.Stat(backupPath); err != nil {
+		os.RemoveAll(backupPath)
+	}
+
+	file, err := os.Open(path)
```
Comment (olegrok) [inline cli/repair/common.go:69]: Please specify Open flags explicitly

File: cli/repair/clusterwide_config.go (hunk 18 above)
Comment (dokshina) [reply cli/repair/clusterwide_config.go:20]: There is no `expelled` key, in case of expelled instance it's config value is `expelled`.
So, `disabled` is an instance config key, `expelled` is a whole instance config value.

E.g.:

```yaml
servers:
  srv-1:
    advertise_uri: localhost:3303
    disabled: true
  drv-2: expelled
```

File: cli/repair/clusterwide_config.go (hunk 11 above)
Comment (dokshina) [reply cli/repair/clusterwide_config.go:119]: If somebody specify null uuid, cli raises an error `no instance with such UUID`. 
The only one non-handled case is: cluster already contains instance with non-correct UUID. This is not `cartridge-cli` business.

There is no possibility to change instance UUID, we only operate existent UUIDs, and we can only validate if cluster contains instance/replicaset with such UUID.

File: cli/repair/repair.go [hunk 20]
```diff
@@ -0,0 +1,95 @@
...
+	"github.com/apex/log"
+	"github.com/tarantool/cartridge-cli/cli/common"
+	"github.com/tarantool/cartridge-cli/cli/context"
+)
+
+type ProcessConfFuncType func(workDir string, ctx *context.Ctx) ([]string, error)
+type PatchConfFuncType func(topologyConf *TopologyConfType, ctx *context.Ctx) error
+
+func List(ctx *context.Ctx) error {
+	ctx.Cli.Verbose = true
```
[the remark this answers is not in the document]
Comment (dokshina) [reply cli/repair/repair.go:16]: Yes, I will fix it in #342

File: cli/commands/repair.go (hunk 17 above)
Comment (dokshina) [reply cli/commands/repair.go:107]: Will be fixed in #339

File: cli/repair/common.go (hunk 19 above)
Comment (dokshina) [reply cli/repair/common.go:69]: Will be fixed in #339

File: cli/repair/clusterwide_config.go (hunk 11 above)
Comment (olegrok) [reply cli/repair/clusterwide_config.go:119]: Well, ok

File: cli/repair/repair.go (hunk 14 above)
Comment (olegrok) [reply cli/repair/repair.go:80]: E.g. "command failed with following messages:"

File: cli/commands/repair.go (hunk 17 above)
Comment (dokshina) [reply cli/commands/repair.go:107]: See [this commit](https://github.com/tarantool/cartridge-cli/pull/339/commits/ab4abb773a9d1d3706bb81f741271f429c08abb4)

File: cli/repair/common.go (hunk 19 above)
Comment (dokshina) [reply cli/repair/common.go:69]: `os.Open` uses `RDONLY` flag by default. I've added file mode flags to created backup file.

See [commit](https://github.com/tarantool/cartridge-cli/pull/339/commits/4ccad749c279cf73dc702d142cf8a3d2e456f7f6)

File: cli/repair/repair.go (hunk 14 above)
Comment (dokshina) [reply cli/repair/repair.go:80]: It can be diff messages in case of `--dry-run`.
